### PR TITLE
Optional, User-sepcified Solvers

### PR DESCRIPTION
**Pull request by scopatz**
This PR provides an interface for users to specify in the input file the solver, preconditioner, and exclusivity of orders in the simulation.  This is 100% optional and defaults maintain backwards compatibility.

### CODE REVIEW

**Comment by scopatz** [comment]
Also, @gidden would ideally review this.  Though @mbmcgarry and @jdangerx (and qual-laden @opotowsky) should also feel empowered to review and merge.

File: share/cyclus-flat.rng.in [hunk 1]
```diff
@@ -18,6 +18,14 @@ datatypeLibrary="http://www.w3.org/2001/XMLSchema-datatypes">
       <optional> 
         <element name="decay"><text/></element> 
       </optional>
+        <element name="solver"> <text/> </element>
+      </optional>
+      <optional>
+        <element name="preconditioner"> <text/> </element>
+      </optional>
+      <optional>
+        <element name="exclusive_orders_only"><data type="boolean" /></element>
```
Comment (gidden) [inline share/cyclus-flat.rng.in]: This flag is a misnomer. It should be named something like `allow_exclusive_orders` or equivalent to match with current implementation.

Comment (gidden) [inline share/cyclus-flat.rng.in]: nevermind, you're correct, I had to double check the implementation

Comment (gidden) [inline share/cyclus-flat.rng.in]: (and documentation)

File: src/sim_init.cc [hunk 2]
```diff
@@ -173,28 +173,68 @@ void SimInit::LoadRecipes() {
...
+  }
+
+  precon = LoadPreconditioner(precon_name);
+
+  if (solver_name == "greedy") {
+    if (precon == NULL) {
+      solver = new GreedySolver(exclusive_orders);
+    } else {
+      solver = new GreedySolver(exclusive_orders, 
+                                (GreedyPreconditioner*) precon);
```
Comment (gidden) [inline src/sim_init.cc]: should be a `static_cast` http://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Casting

**Comment by gidden** [comment]
Generally looks good @scopatz, thanks for tackling this. I noted a single gcsg issue.

Separately, though, I think there should be some documentation related to solver defaults. The best place is likely on our user-specific website (http://fuelcycle.org/user/input_specs/control.html).

**Comment by gidden** [comment]
mac osx error, relaunching

File: src/sim_init.cc (hunk 2 above)
Comment (scopatz) [inline src/sim_init.cc]: Shouldn't it be a `reinterpret_cast`?

**Comment by scopatz** [comment]
Ping! Changed to `reinterpret_cast<>()`

**Comment by scopatz** [comment]
> Separately, though, I think there should be some documentation related to solver defaults. The best place is likely on our user-specific website (http://fuelcycle.org/user/input_specs/control.html).

Yup yup, I am working on adding docs.  I also added a table to the interface.

**Comment by scopatz** [comment]
Looks like this passes CI.

**Comment by gidden** [comment]
I think we perhaps should review the XML structure (and, therefore, back-end table design) given our [listserv](https://groups.google.com/forum/#!topic/cyclus-dev/R1Nd6QYV7Ks) discussion and looking towards the future. Specifically, we need to think about how nesting occurs in our input file. If we move forward with the strategy pattern approach, then a strategy needs to be identifiable in the input file. Because strategies will be specific to the greedy solver, it would make sense to nest the input. Furthermore, other solvers will also have solver-specific (optional) input, such as CLI args for COIN solvers. 

If we move towards nesting, however, we quickly get into an archetype-like pattern, i.e., there are some known solvers, each of which may be uniquely configurable. Note that I'm not advocating for plug-in solvers; I think (for at least the medium or even long term) we should maintain the allowable DRE solvers as part of the kernel. 

I understand that this suggestion will require more work, so perhaps @rwcarlsen can chime in as to his opinion? Should we move this to the listserv?

**Comment by rwcarlsen** [comment]
mailing list definitely.

**Comment by scopatz** [comment]
ping. I have updated the rng to be more in line with the vision @gidden outlined on the mailing list.

**Comment by gidden** [comment]
It looks like our DEVELOP head is failing the test_abi check on my machine, and that is the only test this fails on my machine as well.

**Comment by scopatz** [comment]
That is because #1073 needs to be reviewed and merged.

File: src/sim_init.h [hunk 3]
```diff
@@ -78,6 +78,8 @@ class SimInit {
   void LoadDecomSched();
   void LoadNextIds();
 
+  void* LoadPreconditioner(std::string name);
+  ExchangeSolver* LoadGreedySolver(bool exclusive, std::set<std::string> tables);
```
Comment (gidden) [inline src/sim_init.h]: this line is 80 chars, but just!

Comment (gidden) [inline src/sim_init.h]: sigh, another reason to dislike [gcsg](http://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Line_Length), it accepst <=80 while PEP8 is [<80](https://www.python.org/dev/peps/pep-0008/#maximum-line-length)

Comment (gidden) [inline src/sim_init.h]: (so ignore this)

File: tests/xml_file_loader_tests.cc [hunk 4]
```diff
@@ -71,6 +75,20 @@ std::string XMLFileLoaderTests::ControlSequenceWithDecay() {
...
 }
 
+std::string XMLFileLoaderTests::ControlSequenceWithSolver() {
+  return  "<simulation>"
+          " <control>"
+          "  <duration>1200</duration>"
+          "  <startmonth>1</startmonth>"
+          "  <startyear>2000</startyear>"
+          "  <simstart>0</simstart>"
+          "  <solver>greedy</solver>"
```
Comment (gidden) [inline tests/xml_file_loader_tests.cc]: Perhaps I've misunderstood the schema as defined, but shouldn't this test use a nested block like 
```
<solver>
   <greedy>
       <preconditioner/>
   </greedy>
   <exclusive_orders_only/>
</solver>
```
?

Comment (scopatz) [inline tests/xml_file_loader_tests.cc]: Yes it should.

**Comment by scopatz** [comment]
Test has been updated, good catch @gidden

**Comment by gidden** [comment]
Hmm, do we know why it was passing before?

**Comment by gidden** [comment]
Ok, this looks great, @scopatz, thanks for working with me through the suggestions. I really appreciate you doing such a robust job. I think additional solvers will be much easier to select with this interface.

My final comment is that there is a bit of twice-given information in the `<name>` tag and the name of the solver tag in `<choice>`. We get around this in our archetype reading by having a `<config>` tag wrap the choice, and then inferring the archetype based on what's inside the `<config>` tag. Do you think it's worth using a similar inference pattern here?

**Comment by scopatz** [comment]
It was passing before because the query is based on "/control/solver/name" not "/control/solver" and the fact that when the query comes up negative, the greedy solver is assumed to maintain default backwards compatibility.

**Comment by gidden** [comment]
Ah ok, I guess I assumed our input tests were also validated, which is obviously not true. However, all of cycamore regression tests run fine. Thanks for the edification =).

**Comment by scopatz** [comment]
@gidden, are you suggesting

*A*
```
<optional>
  <element name=config>
    <choice>
      <element name="greedy">
        <interleave>
          <optional>
            <element name="preconditioner"> <text/> </element>
          </optional>
        </interleave>
      </element>
    </choice>
  </element>
</optional>
```

or 

*B*
```
<optional>
  <element name=config>
    <choice>
      <interleave>
        <optional>
          <element name="preconditioner"> <text/> </element>
        </optional>
      </interleave>
    </choice>
  </element>
</optional>
```

**Comment by gidden** [comment]
I am suggesting the following

<optional> <element name="solver"> <interleave> <optional> <config> <choice>
<element name="greedy"> <interleave> <optional> <element
name="preconditioner"> <text/> </element> </optional> </interleave>
</element> </choice> </config> </optional> <optional> <element
name="exclusive_orders_only"> <data type="boolean" /> </element> </optional>
</interleave> </element> </optional>

On Thu, Jan 29, 2015 at 12:20 PM, Anthony Scopatz <notifications@github.com>
wrote:

> @gidden <https://github.com/gidden>, are you suggesting
>
> *A*
>
> <optional>
>   <element name=config>
>     <choice>
>       <element name="greedy">
>         <interleave>
>           <optional>
>             <element name="preconditioner"> <text/> </element>
>           </optional>
>         </interleave>
>       </element>
>     </choice>
>   </element>
> </optional>
>
> or
>
> *B*
>
> <optional>
>   <element name=config>
>     <choice>
>       <interleave>
>         <optional>
>           <element name="preconditioner"> <text/> </element>
>         </optional>
>       </interleave>
>     </choice>
>   </element>
> </optional>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cyclus/cyclus/pull/1070#issuecomment-72076020>.
>



-- 
Matthew Gidden
Ph.D. Candidate, Nuclear Engineering
The University of Wisconsin -- Madison
Ph. 225.892.3192

**Comment by scopatz** [comment]
OK I have gone to the config strategy.

**Comment by gidden** [comment]
this looks perfect, thanks @scopatz. waiting for CI.

**Comment by gidden** [comment]
Hmm, I know that most of CHTC experience a reboot yesterday/today. It's possible we may need to relaunch this.

**Comment by gidden** [comment]
Thanks so much for working through the suggestions both here and on the list serv, @scopatz

**Comment by scopatz** [comment]
np. I am just glad it is over :)

**Pull request merged**